### PR TITLE
[runtime] Allocate the memory for gshared gparams from image sets.

### PR DESCRIPTION
**Pull request by vargaz**
Fixes https://github.com/mono/mono/issues/18127.

### CODE REVIEW

File: mono/mini/mini-generic-sharing.c [hunk 1]
```diff
@@ -3892,12 +3893,15 @@ shared_gparam_equal (gconstpointer ka, gconstpointer kb)
...
 mini_get_shared_gparam (MonoType *t, MonoType *constraint)
 {
+	MonoImageSet *set;
 	MonoGenericParam *par = t->data.generic_param;
 	MonoGSharedGenericParam *copy, key;
 	MonoType *res;
 	MonoImage *image = NULL;
 	char *name;
 
+	set = mono_metadata_merge_image_sets (mono_metadata_get_image_set_for_type (t), mono_metadata_get_image_set_for_type (constraint));
```
Comment (CoffeeFlux) [inline mono/mini/mini-generic-sharing.c]: Don't we need locking around this too?

File: mono/mini/mini-generic-sharing.c [hunk 2]
```diff
@@ -65,6 +65,7 @@ partial_sharing_supported (void)
 {
 	if (!ALLOW_PARTIAL_SHARING)
 		return FALSE;
+	return TRUE;
```
Comment (CoffeeFlux) [inline mono/mini/mini-generic-sharing.c]: Was this intentional? This is going to make all the logic past this dead code.

Comment (vargaz) [reply mono/mini/mini-generic-sharing.c]: Fixed.

File: mono/mini/mini-generic-sharing.c (hunk 1 above)
Comment (vargaz) [reply mono/mini/mini-generic-sharing.c]: The mono_metadata functions take care of that.

**Comment by akoeplinger** [comment]
backport?

**Comment by vargaz** [comment]
@monojenkins backport 2020-02

File: mono/metadata/metadata.c [hunk 3]
```diff
@@ -3363,6 +3375,26 @@ mono_metadata_get_image_set_for_aggregate_modifiers (MonoAggregateModContainer *
...
+mono_metadata_merge_image_sets (MonoImageSet *set1, MonoImageSet *set2)
+{
+	MonoImage **images = g_newa (MonoImage*, set1->nimages + set2->nimages);
+	memcpy (images, set1->images, sizeof (MonoImage*) * set1->nimages);
+
+	int nimages = set1->nimages;
+	// FIXME: Quaratic
+	for (int i = 0; i < set1->nimages; ++i) {
+		int j;
+		for (j = 0; j < set2->nimages; ++j) {
```
Comment (joncham) [inline mono/metadata/metadata.c]: It seems like this merge may be wrong. If `set1` has 1 item and `set2` has 10 items, it seems like we can only ever add one item from `set2` given the add `images [nimages ++] = set2->images [i];` is in the outer loop (based on item count from `set1`).

Comment (CoffeeFlux) [reply mono/metadata/metadata.c]: Yep, oops. I missed that in review, thanks for catching it.

**Pull request merged**